numpy/linalg/tests/test_linalg.py

```diff
@@ -1710,6 +1712,65 @@ def test_mode_all_but_economic(self):
             self.check_qr(m2)
             self.check_qr(m2.T)
 
+    def check_qr_stacked(self, a):
+        # This test expects the argument `a` to be an ndarray or
+        # a subclass of an ndarray of inexact type.
+        a_type = type(a)
+        a_dtype = a.dtype
+        m, n = a.shape[-2:]
+        k = min(m, n)
+
+        # mode == 'complete'
+        q, r = linalg.qr(a, mode='complete')
+        assert_(q.dtype == a_dtype)
+        assert_(r.dtype == a_dtype)
+        assert_(isinstance(q, a_type))
+        assert_(isinstance(r, a_type))
+        assert_(q.shape[-2:] == (m, m))
+        assert_(r.shape[-2:] == (m, n))
+        assert_almost_equal(matmul(q, r), a)
+        assert_almost_equal(swapaxes(q, -1, -2).conj(), np.linalg.inv(q))
+        assert_almost_equal(np.triu(r[..., :, :]), r)
+
+        # mode == 'reduced'
+        q1, r1 = linalg.qr(a, mode='reduced')
+        assert_(q1.dtype == a_dtype)
+        assert_(r1.dtype == a_dtype)
+        assert_(isinstance(q1, a_type))
+        assert_(isinstance(r1, a_type))
+        assert_(q1.shape[-2:] == (m, k))
+        assert_(r1.shape[-2:] == (k, n))
+        assert_almost_equal(matmul(q1, r1), a)
+        assert_almost_equal(swapaxes(q, -1, -2).conj(), np.linalg.inv(q))
+        assert_almost_equal(np.triu(r1[..., :, :]), r1)
+
+        # mode == 'r'
+        r2 = linalg.qr(a, mode='r')
+        assert_(r2.dtype == a_dtype)
+        assert_(isinstance(r2, a_type))
+        assert_almost_equal(r2, r1)
+
+    def test_stacked_inputs(self):
+
+        curr_state = np.random.get_state()
+        np.random.seed(0)
+
+        normal = np.random.normal
+        sizes = [(3, 4), (4, 3), (4, 4), (3, 0), (0, 3)]
+        dts = [np.float32, np.float64, np.complex64]
+        for size in sizes:
+            for dt in dts:
+                a1, a2, a3, a4 = [normal(size=size), normal(size=size),
+                                  normal(size=size), normal(size=size)]
+                b1, b2, b3, b4 = [normal(size=size), normal(size=size),
+                                  normal(size=size), normal(size=size)]
+                A = np.asarray([[a1, a2], [a3, a4]], dtype=dt)
+                B = np.asarray([[b1, b2], [b3, b4]], dtype=dt)
+                self.check_qr_stacked(A)
+                self.check_qr_stacked(B)
+                self.check_qr_stacked(A + 1.j*B)
+
+        np.random.set_state(curr_state)
```

Comment:
I don't think we reset the random number state after tests, that is on each test to seed as it sees fit.

Reply:
Are the seeds globally set for testing purposes? Or in other words can I assume that the tests above would be static in nature?

Reply:
I will remove this setting and resetting of seeds. I noticed a call `np.random.seed(1234)` at the top of the file. So, I believe the seeds are set globally for tests in this file. Thanks for pointing out.

Reply:
Each test that cares is expected to seed the random number generator it uses. I am not sure what you mean by "static in nature". We could at some point consider creating a random number generator fixture, but it seems to me better for any test that cares to call `np.random.seed(xxx)`, which is more verbose but also much clearer.
```

Reply:
Thanks for a detailed overview.

> I am not sure what you mean by "static in nature"

By static tests I meant that are the testing data same for every run of the tests. Specifically, if the seed for each test (globally or locally) is not set then testing may vary everytime we run the tests. However, if the seeds are fixed then the testing won't vary. I observed that `np.random.seed` is called at least once inside the above file and hence, the tests should be static. I was resetting the seeds after my tests because I thought that if there are any seed sensitive tests inside the above file then they shouldn't change. However, I realized that it isn't required, so removed that part.